packages/valory/skills/task_execution/behaviours.py

```diff
@@ -270,10 +299,27 @@ def _check_undelivered_reqs_marketplace(self) -> None:
             counterparty=LEDGER_API_ADDRESS,
             ledger_id=self.context.default_ledger_id,
         )
+        self.params.req_type = "marketplace"
         self.context.outbox.put_message(message=contract_api_msg)
 
     def _execute_task(self) -> None:
         """Execute tasks."""
+        # if no deadline is set it, otherwise continue
+        if self._last_deadline is None:
+            self._last_deadline = self._fetch_deadline()
+
+        # check if the executing task is within deadline or not
+        while self.params.in_flight_req is not False:
+            if time.time() > self._last_deadline:
+                # Deadline reached, restart the task execution
+                self.context.logger.info(
+                    f"Deadline reached for task {self._executing_task}. Restarting task execution..."
+                )
+                self._last_deadline = time.time()
+                self.params.in_flight_req = False
+                self._is_cold_start = False
+                return self._execute_task()
+
         # check if there is a task already executing
         if self.params.in_flight_req:
             # there is an in flight request
```

Comment:
i think it suffices that inside here we have the deadline logic you have added, no need for an additional loop